File: 48_Rotate_Image/0ms.c
void swap(int* a, int* b) {
  int temp = *a;
  *a = *b;
  *b = temp;
}

void reverse(int* arr, int n) {
  int left = 0, right = n - 1;
  while(left < right) {
    swap(&arr[left], &arr[right]);
    left++;
    right--;
  }
}

void rotate(int** matrix, int matrixSize, int* matrixColSize) {
  int n = matrixSize;
  int m = *matrixColSize;
  for(int i = 0; i < n - 1; i++) {
    for(int j = i + 1; j < n; j++) {
      swap(&matrix[i][j], &matrix[j][i]);
    }
  }

  for(int i = 0; i < n; i++) {
    reverse(matrix[i], m);
  }
}

File: 48_Rotate_Image/0ms.java
class Solution {
  public void rotate(int[][] matrix) {
    for(int i = 0; i < matrix.length; i++) {
      for(int j = i + 1; j < matrix.length; j++) {
        int aux = matrix[i][j];
        matrix[i][j] = matrix[j][i];
        matrix[j][i] = aux;
      }
    }

    for(int i = 0; i < matrix.length; i++) {
      for(int j = 0, k = matrix.length - 1; j < matrix.length / 2; j++, k--) {
        int aux = matrix[i][j];
        matrix[i][j] = matrix[i][k];
        matrix[i][k] = aux;
      }
    }
  }
}
